// dc-delivery-sdk-js/build/main/lib/content/coordinators/GetContentItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Url_1 = require("../../utils/Url");
var JsonWalker_1 = require("../../utils/JsonWalker");
var ContentItem_1 = require("../model/ContentItem");
/**
 * @hidden
 */
var LD_ID = '@id';
/**
 * @hidden
 */
var LD_TYPE = '@type';
/**
 * @hidden
 */
var LD_GRAPH = '@graph';
/**
 * @hidden
 */
var GetContentItem = /** @class */ (function () {
    function GetContentItem(config, contentClient, mapper) {
        this.config = config;
        this.contentClient = contentClient;
        this.mapper = mapper;
    }
    GetContentItem.prototype.getContentItem = function (id) {
        var _this = this;
        var url = this.getUrl({
            'sys.iri': "http://content.cms.amplience.com/" + id
        });
        return this.contentClient.get(url).then(function (response) {
            var contentItems = _this.processResponse(response.data);
            if (contentItems.length === 0) {
                return Promise.reject("Content item \"" + id + "\" was not found");
            }
            var item = contentItems.find(function (x) { return x._meta.deliveryId === id; });
            if (!item) {
                return Promise.reject("Content item \"" + id + "\" was not found");
            }
            return Promise.resolve(_this.hydrateContentItem(item));
        });
    };
    GetContentItem.prototype.getUrl = function (query) {
        var args = [
            ['query', JSON.stringify(query)],
            ['fullBodyObject', 'true'],
            ['scope', 'tree'],
            ['store', this.config.account]
        ];
        if (this.config.locale) {
            args.push(['locale', this.config.locale]);
        }
        var queryString = Url_1.encodeQueryString(args);
        return "/cms/content/query?" + queryString;
    };
    /**
     * Convert plain content JSON into ContentItem instamce
     * @param content
     */
    GetContentItem.prototype.hydrateContentItem = function (content) {
        var contentItem = new ContentItem_1.ContentItem();
        contentItem.body = this.mapper.toMappedContent(content);
        return contentItem;
    };
    /**
     * Pre-processes the response to create a single content item object
     * with all linked content items inlined.
     * @param data Response from content query API
     */
    GetContentItem.prototype.processResponse = function (data) {
        var _this = this;
        if (!data.result || !data[LD_GRAPH] || !Array.isArray(data.result)) {
            return [];
        }
        else {
            var graph = data[LD_GRAPH];
            var graphChildrenById_1 = {};
            graph.forEach(function (child) { return (graphChildrenById_1[child[LD_ID]] = child); });
            return data.result.map(function (result) {
                result = graphChildrenById_1[result[LD_ID]];
                // inline linked items
                result = JsonWalker_1.walkAndReplace(result, {
                    beforeWalk: function (node) {
                        if (typeof node === 'object') {
                            if (node[LD_ID]) {
                                node = graphChildrenById_1[node[LD_ID]];
                            }
                        }
                        return node;
                    }
                });
                // upgrade legacy content & remove json-ld
                result = JsonWalker_1.walkAndReplace(result, {
                    beforeWalk: function (node) {
                        node = _this.upgradeLegacyContent(node);
                        node = _this.injectMetaData(node);
                        node = _this.removeJsonLD(node);
                        return node;
                    }
                });
                return result;
            });
        }
    };
    /**
     * Content produced by older versions do not include _meta.schema.
     * This function inserts those values to normalize the output
     */
    GetContentItem.prototype.upgradeLegacyContent = function (node) {
        if (!node) {
            return node;
        }
        var type = node[LD_TYPE];
        var id = node[LD_ID];
        var isImage = id && id.indexOf('http://image.cms.amplience.com/') === 0;
        var isVideo = id && id.indexOf('http://video.cms.amplience.com/') === 0;
        if (type || isImage || isVideo) {
            if (!node._meta) {
                node._meta = {};
            }
            if (type) {
                if (!node._meta.schema) {
                    node._meta.schema = type;
                }
                if (!node._meta.name && node._title) {
                    node._meta.name = node._title;
                }
            }
            else if (isImage) {
                node._meta.schema =
                    'http://bigcontent.io/cms/schema/v1/core#/definitions/image-link';
            }
            else if (isVideo) {
                node._meta.schema =
                    'http://bigcontent.io/cms/schema/v1/core#/definitions/video-link';
            }
        }
        return node;
    };
    /**
     * Injects additional meta data which is lost by removing
     * the JSON-LD properties
     */
    GetContentItem.prototype.injectMetaData = function (node) {
        if (node) {
            var id = node[LD_ID];
            if (id) {
                if (id.indexOf('http://content.cms.amplience.com/') === 0) {
                    node._meta = node._meta || {};
                    node._meta.deliveryId = id.slice(33);
                }
                else if (id.indexOf('http://image.cms.amplience.com/') === 0 ||
                    id.indexOf('http://video.cms.amplience.com/') === 0) {
                    node.id = id.slice(31);
                }
            }
        }
        return node;
    };
    /**
     * JSON-LD keywords are just for delivery payload
     */
    GetContentItem.prototype.removeJsonLD = function (node) {
        if (node) {
            delete node[LD_TYPE];
            delete node[LD_ID];
        }
        return node;
    };
    return GetContentItem;
}());
exports.GetContentItem = GetContentItem;
